### ISSUE TITLE
[Question] Low accuracy in VAD when inference?

### ISSUE THREAD

**Issue by guyqaz** [NVIDIA/NeMo#2353]
I trained VAD model using my own dataset follow instructions in [tutorial 6](https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/06_Voice_Activiy_Detection.ipynb)
My dataset is Vietnamese telephony conversation, 15 hours with 8 hours of speech and 7 hours of noise. I got accuracy 95% on my test set.
The format of dataset quite similar with google speech dataset. Each sample of speech is one statement and have duration 2s to 5s.

When i follow instruction in [tutorial 7](https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/07_Online_Offline_Microphone_VAD_Demo.ipynb) to inference sample in test set what contain only one statement, the prediction is good. The problem is when i inference long audio has duration 20s of speech and noise, the prediction is really bad.

This is predict sample of speech has only one statement
![foo](https://user-images.githubusercontent.com/83386486/121889898-f44b2380-cd43-11eb-9a6f-6b5abf18b1e9.png)

And long audio, 0 to 5s of audio is noise.
![bar](https://user-images.githubusercontent.com/83386486/121889893-f2816000-cd43-11eb-8dc0-929b4f2af029.png)

1. Any idea to deal with this problem
2. I want to tune threshold but it required RTTM ground truth, what is the format of RTTM file and how to generate this file from dataset like google speech.
3. How to tune best window size with my own dataset

**Comment by fayejf** [comment]
hi @guyqaz 
Some questions have been answered here 
https://github.com/NVIDIA/NeMo/issues/1632#issuecomment-861042671

For some questions in https://github.com/NVIDIA/NeMo/issues/1632 which are related here:

>  The sample rate of long audio sample is 16000Hz.

Then this is odd to see there is no prediction after 15s in figure 2. 

 
> Now i am defining one audio of sentence as speech, not a word level like google speech dataset. But i still using process_vad_data to generate manifest file.
> When infer audio just has statement or noise, i think it good because almost all frame are predicted to be speech or all noise. But when predict long audio have both speech and noise. Model dose not seen recognize speech and noise.


I think this should be issue. MarbleNet expects the input segment to be a homogeneous speech/non-speech, even though it is capable to learn with some noisy data. Maybe directly using process_vad_data for audio of sentence to generate segments raises a lot mislabeled training data.

> My dataset also have label with shorter duration what label for stop point, noise in audio contain speech of statement.

I'm not quite understand this sentence. Could you explain? 

> Do you think i should using shorter label or just using audio of sentence is speech.

I don't think using `audio of sentence as speech` would work cause it might contain lot of mislabelled noise.


> Is there a tutorial to tune threshold from output of predict frame level vad_infer and manifest file.

I believe use RTTM format as groundtruth is a easier way and you can use this [vad_tune_threshold script](https://github.com/NVIDIA/NeMo/blob/main/scripts/voice_activity_detection/vad_tune_threshold.py) directly

frame level prediction (w/ or wo/posterior) -> to convert to speech / no-speech given threshold -> compare with ground truth label to find optimal threshold 

If you have frame level **label**, for example, [1,1,1,1,1,0,0,1,1], 
and your frame level **prediction probability** is like [0.8, 0.9, 0.7, 0.6, 0.5, 0.4, 0, 0.2, 0.7]
if prob > threshold, then pred is 1 else 0. And use metrics such as Accuracy to find which threshold gives best result. 

We decide to support threshold tuning with RTTM label for now and might add more helper scripts for different type of label files. It will be much appreciated to add your contribution to NeMo.  😄

**Comment by guyqaz** [comment]
Thanks a lot for your response. 

> I think this should be issue. MarbleNet expects the input segment to be a homogeneous speech/non-speech, even though it is capable to learn with some noisy data. Maybe directly using process_vad_data for audio of sentence to generate segments raises a lot mislabeled training data.

I solved this. The problem as you say. Use statement as speech make my training data has a lot miss label.


> We decide to support threshold tuning with RTTM label for now and might add more helper scripts for different type of label files. It will be much appreciated to add your contribution to NeMo. 

I use logits and label in eval part of tutorial/asr/06.. to tune threshold. Its much simpler way

My model got 97% accuracy when predict frame level on test set. I think it quite good and trying to deploy trained model now.

In [Tutorial 07](https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/07_Online_Offline_Microphone_VAD_Demo.ipynb)

> It is not a recommended way to do inference in production workflows. If you are interested in production-level inference using NeMo ASR models, please sign-up to Jarvis early access program: https://developer.nvidia.com/nvidia-jarvis

Do you have any solution to deploy real time VAD model in very large scale of users . I  want to get signal of speech and then pass it into ASR model.

Again, thank you so much!!

**Comment by guyqaz** [comment]
Hi @fayejf .
> What is the best solution to deploy real time VAD model in very large scale of users? After VAD model, I will pass speech signal to ASR model.

Is this Javis?
I see in the [Javis documentaion](https://docs.nvidia.com/deeplearning/jarvis/user-guide/docs/quick-start-guide.html)

> Models Available for Deployment: When using either of the push-button deployment options, Jarvis uses pre-trained models from NGC. You can also fine-tune custom models with NVIDIA NeMo. Creating a model repository using a fine-tuned model trained in NeMo is a more advanced approach to generating a model repository.

One more thing, I fine-tune custom models with NVIDIA NeMo but save checkpoint with .ckpt format.
Is it possible to deploy .ckpt format model with javis?

Thank for your help?

**Comment by fayejf** [comment]
@guyqaz  Hey


> What is the best solution to deploy real time VAD model in very large scale of users? After VAD model, I will pass speech signal to ASR model.

Jarvis is the recommended production-level inference but currently it does not have VAD model integrated but you are welcome to use your pre-trained VAD model with Jarvis. Talking about ASR, **Jarvis supports good ASR models in multiple languages**, and well-maintained explicit inference pipeline

And you can always convert the model to .onnx 😄 


> Is it possible to deploy .ckpt format model with Jarvis?

I think so, there are some cool tools created from Jarvis team could help you use nemo ckpt in Jarvis

**Comment by guyqaz** [comment]
Hi @fayejf , 
I checked and realized that exp_manager of NeMo also save best checkpoint with postfix = ".nemo"

> If you trained your model with the recent NeMo release (1.0.0.b4), you can directly use tlt … export from

Nvidia Transfer Learning Toolkit provides script to convert .nemo model to .ejrvs format for building and deploying with Jarvis ServiceMaker. I will try it.

Anyway, thanks you for great job. You helped me a lot!

I am closing this issue here.

**Comment by fayejf** [comment]
@guyqaz  Thank you! Enjoy! 😆
